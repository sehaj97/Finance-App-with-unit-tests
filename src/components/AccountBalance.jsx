import React, { useState, useEffect } from "react";
import Notification from "./Notification.jsx";

const AccountBalance = (props) => {
  const [balance, setBalance] = useState();
  const [savingsBalance, setSavingsBalance] = useState();

  useEffect(() => {
    setBalance(props.accounts.balance);
    setSavingsBalance(props.accounts.savingsBalance);
  }, [props.accounts.balance, props.accounts.savingsBalance]);

  const handleSavings = () => {
    if (balance > 1000) {
      setBalance(balance - 100);
      setSavingsBalance(savingsBalance + 100);
    }
  };

  const handleSpending = () => {
    setBalance(balance + 100);
    setSavingsBalance(savingsBalance - 100);
  };

  return (
    <div className="columns">
      <div className="column">
        <div className="box">
          <h4 className="title is-4">Your account balance:</h4>
          <div className="amount balance">{balance}$</div>
          <button
            id="balance-button"
            className="button is-info"
            onClick={() => handleSavings()}
          >
            Send 100$
          </button>
        </div>
      </div>
      <div className="column">
        <div className="box">
          <h4 className="title is-4">Your savings balance:</h4>
          <div className="amount savings">{savingsBalance}$</div>
          <button className="button is-info" onClick={() => handleSpending()}>
            Send 100$
          </button>
        </div>
      </div>
      <Notification balance={balance} />
    </div>
  );
};

export default AccountBalance;
