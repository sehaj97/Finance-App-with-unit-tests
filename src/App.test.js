import React from "react";
import { shallow, mount } from "enzyme";
import App from "./App";
import AccountBalance from "./components/AccountBalance";
import Notification from "./components/Notification";
import toJson from "enzyme-to-json";

const userBalance = {
  balance: 1100,
  savingsBalance: 103,
};

// unit test
it("1. renders App without crashing", () => {
  shallow(<App />);
});

// notification component test
describe("renders notification component", () => {
  it("renders App without crashing", () => {
    shallow(<Notification />);
  });
});

// App component test
describe("renders component", () => {
  it("renders App without crashing", () => {
    shallow(<App />);
  });

  it("renders header without crashing", () => {
    const wrapper = shallow(<App />);
    const welcomeHeader = (
      <h1 className="has-text-centered title is-1">
        Welcome in the personal finance app!
      </h1>
    );
    expect(wrapper.contains(welcomeHeader)).toEqual(true);
  });
  it("renders button", () => {
    const wrapper = mount(<AccountBalance accounts={userBalance} />);
    const label = wrapper.find("#balance-button").text();
    expect(label).toEqual("Send 100$");
  });
});

// test for props
describe("passing props properly", () => {
  const accountWrapper = mount(<AccountBalance accounts={userBalance} />);
  const notificationWrapper = mount(
    <Notification balance={userBalance.balance} />
  );
  it("accepts user account props", () => {
    expect(accountWrapper.props().accounts).toEqual(userBalance);
  });
  it("contains savingBalance value", () => {
    const value = accountWrapper.find(".savings").text();
    const expectedValue = userBalance.savingsBalance + "$";
    expect(value).toEqual(expectedValue);
  });
  it("notification accepts props", () => {
    expect(notificationWrapper.props().balance).toEqual(userBalance.balance);
  });
});

// test logic
describe("logic", () => {
  const wrapper = mount(<AccountBalance accounts={userBalance} />);
  const notificationWrapper = mount(
    <Notification balance={userBalance.balance} />
  );
  wrapper.find("#balance-button").simulate("click");
  it("button click - update savings", () => {
    const savingsValue = wrapper.find(".savings").text();
    const expectedValue = userBalance.savingsBalance + 100 + "$";
    expect(savingsValue).toEqual(expectedValue);
  });
  it("button click - update balance", () => {
    const balanceValue = wrapper.find(".balance").text();
    const expectedBalanceValue = userBalance.balance - 100 + "$";
    expect(balanceValue).toEqual(expectedBalanceValue);
  });
});

describe("snapshots", () => {
  it("App snapshot", () => {
    const tree = shallow(<App />);
    expect(toJson(tree)).toMatchSnapshot();
  });
  it("Accounts snapshots", () => {
    const accountBalanceTree = shallow(
      <AccountBalance accounts={userBalance} />
    );
    expect(toJson(accountBalanceTree)).toMatchSnapshot();
  });
  it("Notification snapshot", () => {
    const notificationTree = shallow(<Notification />);
    expect(toJson(notificationTree)).toMatchSnapshot();
  });
});
