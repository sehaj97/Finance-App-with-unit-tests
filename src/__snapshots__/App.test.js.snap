// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots Accounts snapshots 1`] = `
<div
  className="columns"
>
  <div
    className="column"
  >
    <div
      className="box"
    >
      <h4
        className="title is-4"
      >
        Your account balance:
      </h4>
      <div
        className="amount balance"
      >
        $
      </div>
      <button
        className="button is-info"
        id="balance-button"
        onClick={[Function]}
      >
        Send 100$
      </button>
    </div>
  </div>
  <div
    className="column"
  >
    <div
      className="box"
    >
      <h4
        className="title is-4"
      >
        Your savings balance:
      </h4>
      <div
        className="amount savings"
      >
        $
      </div>
      <button
        className="button is-info"
        onClick={[Function]}
      >
        Send 100$
      </button>
    </div>
  </div>
  <Notification />
</div>
`;

exports[`snapshots App snapshot 1`] = `
<div
  className="app container"
>
  <div
    className="section"
  >
    <h1
      className="has-text-centered title is-1"
    >
      Welcome in the personal finance app!
    </h1>
    <AccountBalance
      accounts={
        Object {
          "balance": 1500,
          "savingBalance": 1328,
        }
      }
    />
  </div>
</div>
`;

exports[`snapshots Notification snapshot 1`] = `
<div
  className="notification is-danger hidden"
>
  <button
    className="delete"
    onClick={[Function]}
  />
  <p>
    Your account balance is very low.
  </p>
  <p>
    You can't transfer more money today.
  </p>
</div>
`;
